[1.12] Allow enchantments to modify damage based on the entity hit

maxanier opened the pull request:
## Why
Enchantments are currently able to make enchanted weapons deal more damaged based on the hit entity's `EnumCreatureAttribute`.
However, it might be interesting to change this based on the entity's attributes. For example I would like to create an enchantment that only deals extra damage do vampire players, therefore I need to check the player entity's capabilities. 
I've tried to use the `Enchantment#onEntityDamaged` (attack the player again in there) for this, but this does not really work as some of the damage calculation is ignored then. Also it is called twice after attacking somehow.

## What
This PR adds an additional method (`calcDamageByEntity`), which calls the 'old' (`calcDamageByCreature`) by default, to the Enchantment class.
`EnchantmentHelper` calls the new method instead of the old one now.
Unfortunately a few more classes had be modified to pass the `Entity` parameter instead of the `EnumCreatureAttribute`.

The entity parameter can be null (ItemStacks use this to generate a damage value for the description). I've added a `Nullable` annotation to make this clear.

## Addendum
I was not sure if I should deprecate the `calcDamageByCreature` method as it might still be interesting for modders to override, but should not be called. (While writing this I decided that it would probably be a good idea)

Also not sure about commenting in patches, so I only added ones where they are necessary.

maxanier committed:
Added a entity sensitive damage calculation to enchantments

liach commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -12,7 +12,22 @@
++    /**
++     * Consider using Forge's entity sensitive version
++     */
+     public float func_152376_a(int p_152376_1_, EnumCreatureAttribute p_152376_2_)
```
Should probably deprecate this one instead.

liach commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -12,7 +12,22 @@
+         return 0;
+     }
+ 
++    public float calcDamageByEntity(int level, @javax.annotation.Nullable Entity entity)
```
You should move this to a forge block in the end of the class.

maxanier replied:
Wanted to keep the patch small, but does make stuff cleaner. Done.

liach commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,6 +1,22 @@
+         return field_77520_b.field_77497_a;
+     }
+ 
+-    public static float func_152377_a(ItemStack p_152377_0_, EnumCreatureAttribute p_152377_1_)
```
Probably leaving the old one and deprecating it is better.

maxanier replied:
Should the new method be moved to a Forge block or stay next to this? :D

liach replied:
Probably leave it before subclasses and after all methods (in a forge block is better)

liach commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -27,3 +43,21 @@
+         {
+             public float field_77495_a;
+-            public EnumCreatureAttribute field_77494_b;
++            public @javax.annotation.Nullable Entity field_77494_b;
```
Call this entity... Or there may be confusions for mappings.

maxanier committed:
Added a entity sensitive damage calculation to enchantments

maxanier committed:
Keep old EnchantmentHelper method

maxanier commented:
Updated based on liach's suggestions (thx).

liach commented:
This looks great to me.

maxanier commented:
@mezz Any concerns or is this good to go?

mezz commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,6 +1,25 @@
+-        field_77521_c.field_77494_b = p_152377_1_;
+-        func_77518_a(field_77521_c, p_152377_0_);
+-        return field_77521_c.field_77495_a;
++        return 0.0F;
```
I'm not sure I see the purpose of changing this.
Deprecating it is fine, but why return 0?

maxanier replied:
Unobfuscated old code:
``` java
        ENCHANTMENT_MODIFIER_LIVING.livingModifier = 0.0F;
        ENCHANTMENT_MODIFIER_LIVING.entityLiving = creatureAttribute;
        applyEnchantmentModifier(ENCHANTMENT_MODIFIER_LIVING, stack);
        return ENCHANTMENT_MODIFIER_LIVING.livingModifier;
```

The old code is not working anymore since the ENCHANTMENT_MODIFIER_LIVING now requires an Entity instead of an EnumCreatureAttribute to calculate it's value. Since this is not available in this context I decided to just return 0 in this deprecated method.

This might actually cause minor problems with mods that do use this method directly instead of indirectly via EntityPlayer or EntityMob as those mods would not find out the actual value for some enchantments.

However, the only other (non breaking) solution I can think of, is to create second ENCHANTMENT_MODIFIER_(CREATURE_OLD) (or integrate it in the current one) which can make use of a EntityCreatureAttribute. But since this would ignore any values added to the enchantments in newer ways this would be a somewhat ugly solution. It would solve most of the issues through.

*EDIT:* After writing this, I think, using the ugly solution might be better than not using it.
The issue of "old" mods not getting the correct value for all enchantments is not that problematic, but
without a breaking change it is kinda hard to notice. 
So this would allow old mods to get the correct value for the old type (includes vanilla) enchantments at least.

 *Usage*     | *Enchantment*   | *Works*              |
|-------------|-----------------|----------------------|
| Vanilla     | Old and New     | Yes                  |
| Updated Mod | Old and New     | Yes                  |
| Old Mod     | Old and Vanilla | Only with workaround |
| Old Mod     | New             | No                   |

mezz replied:
I think we should try to support old mods without breaking them.

maxanier committed:
Added (partial) compatibility for older mods

maxanier committed:
Resolve conflicts

maxanier commented:
Rebased to resolve conflicts and adjusted so that if mods call the old method, so only an EnumEntityAttribute is available, the old Enchantment#calcDamageByCreature is used. Which should get the correct result for vanilla and any mods using the old method.

liach commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -48,3 +59,20 @@
+         }
+     }
++
++    /* ======================================== FORGE START =====================================*/
```
Can you move 
````patch
+    /**
+     * This applies specifically to applying at the enchanting table. The other method {@link #canApply(ItemStack)}
+     * applies for <i>all possible</i> enchantments.
+     * @param stack
+     * @return
+     */
+    public boolean canApplyAtEnchantingTable(ItemStack stack)
+    {
+        return stack.func_77973_b().canApplyAtEnchantingTable(stack, this);
+    }
+
+    /**
+     * Is this enchantment allowed to be enchanted on books via Enchantment Table
+     * @return false to disable the vanilla feature
+     */
+    public boolean isAllowedOnBooks()
+    {
+        return true;
+    }
````
to the forge block?

howtonotwin replied:
You're missing a space at the end of this. Should be like:

    /* ============ FORGE START ============ */

not

    /* ============ FORGE START ============*/

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -12,7 +12,18 @@
+ 
++    /**
++     * Consider using Forge's entity sensitive version
++     */
```
Do __*not*__ add Javadoc to Vanilla methods. Their Javadoc is part of the MCP mappings and should not be modified by Forge. The convention is to just stick a normal comment after `@Deprecated`.

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,6 +1,23 @@
++    /**
++     * Does nothing
++     * Use {@link EnchantmentHelper#getModifierForCreature(ItemStack, Entity)} instead
++     */
```
Same as above: do __*not*__ modify Vanilla methods' Javadoc.

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -48,3 +59,20 @@
++
++    /**
++     * Calculate the damage this enchantment adds to an attack on the given entity
++     */
```
Add a period and

```
@param level the level of this enchantment on the item doing the attack
@param entity the entity being attacked
@return the additional damage to do
```

to the Javadoc (preserving case exactly, so as to conform to the convention).

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -27,3 +44,42 @@
+ 
++    /* ======================================== FORGE START =====================================*/
++
++    public static float getModifierForCreature(ItemStack stack, @javax.annotation.Nullable Entity entity)
```
Javadoc this.

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -27,3 +44,42 @@
+             public void func_77493_a(Enchantment p_77493_1_, int p_77493_2_)
+             {
+-                this.field_77495_a += p_77493_1_.func_152376_a(p_77493_2_, this.field_77494_b);
++                this.field_77495_a += attribute == null? p_77493_1_.calcDamageByEntity(p_77493_2_, this.entity):p_77493_1_.func_152376_a(p_77493_2_,attribute);//Check attribute instead of entity if set for compatibility
```
Ternaries should be spaced out, as `cond ? tbr : fbr`, not `cond?tbr:fbr`.

howtonotwin commented on patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch:
```diff
@@ -273,7 +273,23 @@
+-                else
+-                {
+-                    f1 = EnchantmentHelper.func_152377_a(this.func_184614_ca(), EnumCreatureAttribute.UNDEFINED);
+-                }
```
Unnecessary patching. The next line sets `f1` too, so there's no need to remove this `if`/`else`.

maxanier replied:
sounds strange to do some unnecessary calculations, but it decreases patch size, so ok.

covers1624 replied:
Or just whack an if (false) at the top of the if statement.

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -48,3 +59,20 @@
++     */
++    public float calcDamageByEntity(int level, @javax.annotation.Nullable Entity entity)
++    {
++        return func_152376_a(level,entity instanceof EntityLivingBase?((EntityLivingBase) entity).func_70668_bt():EnumCreatureAttribute.UNDEFINED);
```
Ternaries should be spaced out, as `cond ? tbr : fbr`, not `cond?tbr:fbr`.

maxanier committed:
Updated style

maxanier committed:
Updated style II

maxanier commented:
Thx for the feedback. Applied your suggestions

howtonotwin commented on patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch:
```diff
@@ -21,10 +29,21 @@
          return false;
      }
  
++
```
Typo?

maxanier replied:
Have missed that when moving the other methods in the Forge block. Thx

maxanier committed:
Removed unnecessary newline

maxanier commented:
Not sure why the licence/cla check fails as I have signed it before (and again now).

But besides that, any other concerns @mezz?
"Old" mods can now get the correct enchantment modifier for all enchantments except modded ones which utilize the new method.

mezz commented:
Bump, unstale. 
Looks like the CLA check is passing now.
I'll take a look at this soon.

mezz commented:
Looking good. Please add a test mod so I can try this out, and also to serve as an example usage of this new feature.

maxanier committed:
Added testmod and fixed entity attribute not being reset

maxanier commented:
Rebased onto latest 1.12.x branch to make sure it still works
Fixed another issue
Added Testmod

From EnchantmentDamageTest:
```
/**
 * Test for calculating enchantment damage modifier based on the entity object instead of the creature attribute.
 * <p>
 * HowTo Test:
 * <p>
 * a) Grab yourself a wooden sword and enchant it with "enchantmentdamagetest.new_enchantment"
 * Spawn a Zombie and any other animal (e.g. a cow)
 * The Zombie should be one hit with this sword, the cow should take multiple hits (as normal)
 * <p>
 * b) To verify that the old (deprecated/vanilla) way/method still works
 * enchant a wooden sword with "enchantmentdamagetest.old_enchantment"
 * and repeat the test with a undead creature (Zombie) and a non undead creature (cow).
 * It should have the same result
 */
```

mezz commented:
Tested and looks good to me.

LexManos commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -27,3 +43,49 @@
+         {
+             public float field_77495_a;
+-            public EnumCreatureAttribute field_77494_b;
++            public @javax.annotation.Nullable Entity entity;
```
Binary breaking change, we can't allow this...

maxanier commented:
Mhh, good point, not sure why I changed the name of the attribute anyway, has been some time.

Maybe because I wanted to add a @Nullable annotation and wasn't sure how this would affect de/obfuscation.

I will rename the attribute to the old name. Should I leave the annotation or remove it as well?

mezz commented:
The name change is in a package-private class so nothing can really reference it or break when it changes.

LexManos commented:
You know better then that Mezz, modders access all kinda of dumb things.

mezz commented:
@maxanier you can leave the annotation, it is fine. Please update this to resolve merge conflicts as well.

maxanier committed:
Reverted attribute rename, resolved merge conflict updated test description

maxanier committed:
Actually fixed merge conflict

maxanier commented:
Rebased and resolved conflicts as well as renamed the attribute to the original name to avoid binary breaking changes.

mezz commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -27,3 +42,49 @@
+             public float field_77495_a;
+-            public EnumCreatureAttribute field_77494_b;
++            public @javax.annotation.Nullable Entity entity;
++            public @javax.annotation.Nullable EnumCreatureAttribute field_77494_b;
```
`Nullable` should go on a new line before the `EnumCreatureAttribute` field, so the original line is not edited.
the new `entity` field should go after the vanilla fields

liach replied:
Hmm, this field should be `EnumCreatureAttribute.UNDEFINED` by default instead of `null`.

mezz replied:
That's how mojang has it (null)... the way the class itself is used is very strange.

maxanier replied:
Will change that.

Yeah, this has one of the strangest concepts I have seen in Minecraft so far

mezz commented on patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,6 +1,21 @@
+         return field_77520_b.field_77497_a;
+     }
+ 
++
```
This newline should not be added

maxanier replied:
Sorry, will remove that
My git diff highlighting is always confused by all the +- in the patch files :angry:

mezz commented on patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch:
```diff
@@ -287,6 +287,15 @@
+                 {
+                     f1 = EnchantmentHelper.func_152377_a(this.func_184614_ca(), EnumCreatureAttribute.UNDEFINED);
+                 }
+-
```
leave this newline here, do not remove it

maxanier replied:
~~Why? Idea was to prevent an additional line as described in the Contribution Guidelines https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge#conventions-for-coding-patches-for-a-minecraft-class-javapatch~~

liach replied:
that's to avoid patches, but here you are making patches.

maxanier replied:
I'm sorry, I don't get this. Aren't you always making patches if you modify/patch Minecraft code?
Fixed anyway

maxanier replied:
Never mind, I think I have misinterpreted the guidelines, sorry

mezz reviewed (changes requested):
looks good, just needs some minor fixes for the patches

maxanier committed:
Updated patch style

maxanier commented:
Applied requested changes

LexManos commented:
Gunna need to talk to Grum about the lack of thread safety in EnchantmentHelper.... 
Why are you hijacking ENCHANTMENT_MODIFIER_LIVING? Why not just make your own modifier? Then you wouldn't have to hack all the dual fields stuff...
Also your modification to EntityPlayer is just... wrong... At first glance it looks like you're doubling up the modifier. Which you are in a sense. But you're also applying it to the wrong value. 
You're applying the value to the draw damage. Not the enchanted value that is then multiplied by the cooldown strength. You bypass that whole mechanic....

maxanier commented:
Thanks for noticing the wrong calculation. Sorry about that. I looks like I somehow accidentally changed the `f1` to `f`. Not sure how I was able to screw that up as it was just copy and paste.

Did not create my own modifier because I wanted to keep the changes as minimal as possible (and in one of the earlier version I just replaced the entityAttribute attribute with the entity object as far as I can remember so there was less reason to create my own version.

maxanier commented:
@LexManos did you have a chance to talk with Grum about the thread safety?

maxanier commented:
Any news about this? Kind of stale 

Will fix the conflicts once Lex decides if it will be merged (in the current implementation)

mezz commented:
Waiting on reply from @LexManos

LexManos commented:
I haven't talked to Grum about the thread safety beyond preliminary 'things are broken'. So there really isn't anything we can do for it for now.

liach commented:
There is a few merge conflicts.

maxanier commented:
I won't update this before there is some new information about thread safety from Lex or Grum

MajorTuvok commented:
This just needs some thread saftey comments...
Go away stale, this is a nice PR...

liach commented:
as long as entities are read but not modified (which happens most of the time), things should be fine
